applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <gpio.h>
+
+#include "selector_hw_def.h"
+#ifndef CONFIG_DESKTOP_SELECTOR_HW_MAX_STATES
+#error "Please provide a value for CONFIG_DESKTOP_SELECTOR_HW_MAX_STATES"
+#endif /* CONFIG_DESKTOP_SELECTOR_HW_MAX_STATES */
+
+#include "event_manager.h"
+#include "selector_event.h"
+#include "power_event.h"
+
+#define MODULE selector_hw
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+#define DEFAULT_STATE ARRAY_SIZE(pins)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+	STATE_ERROR,
+};
+
+struct selector {
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t state;
+	u8_t id;
+	u8_t pin_cnt;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(pins)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	BUILD_ASSERT_MSG(ARRAY_SIZE(pins) <= UCHAR_MAX,
+			 "Maximum number of selectors is 255");
+	__ASSERT_NO_MSG(selector->state < ARRAY_SIZE(pins[selector->id]));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_src_type = SELECTOR_SRC_TYPE_HW;
+	event->selector_src_id = selector->id;
+	event->state = selector->state;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value = 0;
+	const struct gpio_pin *sel_pins = pins[selector->id];
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+					sel_pins[i].pin,
+					&value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			return err;
+		}
+
+		if (value) {
+			selector->state = i;
+			break;
+		}
+	}
+
+	if (!value) {
+		selector->state = DEFAULT_STATE;
+		LOG_WRN("Selector %" PRIu8 " state is unknown.", selector->id);
+	}
+	selector_event_send(selector);
+
+	return 0;
+}
+
+static int enable_interrupts_lock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = pins[selector->id];
+	int key = irq_lock();
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		if (i == selector->state) {
+			continue;
+		}
+
+		int err = gpio_pin_enable_callback(gpio_dev[sel_pins[i].port],
+						   sel_pins[i].pin);
+		if (err) {
+			LOG_ERR("Cannot enable callback (err %d)", err);
+			return err;
```

Comment:
oops!

Reply:
Added `irq_unlock(key);` Thanks

Reply:
I would prefer break and single exit point

Reply:
ok